server/app/auth/ApiAuthenticator.java

```diff
@@ -0,0 +1,116 @@
+ * reach a controller are already authenticated, but it is the controller's responsibility to check
+ * that the resource being accessed is authorized for the request's API key.
+ *
+ * <p>The API key ID and secret are provided as the basic auth username and password, respectively.
+ * To be valid, a request must reference a valid API key with its username credential, and further
+ * that key must:
+ *
+ * <ul>
+ *   <li>Not be retired.
+ *   <li>Have an expiration date in the future.
+ *   <li>Have a subnet that includes the IP address of the request.
+ *   <li>Have a salted key secret that matches the salted password in the request's basic auth
+ *       credentials.
+ * </ul>
+ */
+public class ApiAuthenticator implements Authenticator {
+
+  // The cache expiration time is intended to be long enough reduce database queries from
+  // authenticating API calls while being short enough that if an admin retires a key or
+  // otherwise edits it, the edits will take effect within a reasonable amount of time.
+  // When tuning this value, consider the use-case of an API consumer rapidly paging through
+  // a list API, and also consider the admin retiring a key when they've discovered it has
+  // been compromised.
+  private static final int CACHE_EXPIRATION_TIME_SECONDS = 60;
+
+  private final SyncCacheApi syncCacheApi;
+  private final Provider<ApiKeyService> apiKeyService;
+
+  @Inject
+  public ApiAuthenticator(
+      @NamedCache("api-keys") SyncCacheApi syncCacheApi, Provider<ApiKeyService> apiKeyService) {
+    this.syncCacheApi = checkNotNull(syncCacheApi);
+    this.apiKeyService = checkNotNull(apiKeyService);
+  }
+
+  /**
+   * Authenticates that an API request has a valid API key and is from a permitted IP address.
+   * Throws a {@link BadCredentialsException} if not, causing a status-only HTTP response 401. The
+   * exception messages are included in the server logs to aid in debugging and monitoring for
+   * malicious use.
+   */
+  @Override
+  public void validate(Credentials rawCredentials, WebContext context, SessionStore sessionStore) {
+    if (!(rawCredentials instanceof UsernamePasswordCredentials)) {
+      throw new RuntimeException("ApiAuthenticator must receive UsernamePasswordCredentials.");
+    }
+
+    UsernamePasswordCredentials credentials = (UsernamePasswordCredentials) rawCredentials;
+    String keyId = credentials.getUsername();
+
+    // Cache the API key for quick lookup in the controller, also for subsequent requests.
+    // We intentionally cache the empty optional rather than throwing here so that subsequent
+    // requests with the invalid key do not put pressure on the database.
+    ApiKey apiKey =
+        syncCacheApi
+            .getOrElseUpdate(
+                keyId, () -> apiKeyService.get().findByKeyId(keyId), CACHE_EXPIRATION_TIME_SECONDS)
+            .orElseThrow(() -> new BadCredentialsException("API key does not exist"));
+
+    if (apiKey.isRetired()) {
```

Comment:
? Per the discussion you had with Kiana, would it be appropriate here to comment that the exception messages don't make it to the user?  

IE: just making it clear that this isn't informative to the user, or indicative of the particular failure to them.

Reply:
added a paragraph to the end of the class doc